### PR TITLE
Laboratory work #3, Ratmir Akhmatzyanov - 22FPL1 

### PR DESCRIPTION
**Pull request by Rotmerka**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_3_generate_by_ngrams/target_score.txt [hunk 1]
```diff
@@ -1 +1 @@
-0
+4
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/target_score.txt:1]: I guess it is not enough

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -42,6 +44,12 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case any of methods used return None, None is returned.
         """
 
+        if not isinstance(text, str):
+            return None
+
+        lower_list_text = tuple(filter(lambda x: x == " " or str.isalpha, text.lower()))
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:50]: `x == " "` -> not x

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -76,6 +88,10 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+        position = list(self._storage.values()).index(element_id)
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:93]: tooooo dificult

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
The same comments

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -129,23 +123,10 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
-        if not (isinstance(text, str) and text):
-            return None
 
-        tokenized_text = self._tokenize(text)
-        if not tokenized_text:
+        if not (isinstance(text, str) and text):
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:127]: where is the implementation?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -251,15 +228,10 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+        if not (isinstance(decoded_corpus, tuple) and decoded_corpus):
             return None
 
-        decoded_text = ''.join(decoded_corpus).replace('_', ' ').capitalize()
-
-        if decoded_text[-1] == ' ':
-            return f"{decoded_text[:-1]}."
-
-        return f"{decoded_text}."
+        return f"{''.join(decoded_corpus).replace(self._end_of_word_token, ' ').capitalize().rstrip()}."
```
[the remark this answers is not in the document]
Comment (Rotmerka) [reply lab_3_generate_by_ngrams/main.py]: like that?

**Pull request closed without merging**